File: python/tools/list_files.py
"""
List files tool implementation.
Provides directory listing functionality with schema caching.
"""

import json
import os
from typing import Any, Dict, Optional
from agent import ToolDefinition


class ListFilesTool(ToolDefinition):
    """
    Tool for listing files and directories.
    Provides directory listing functionality with cached schema for performance.
    """

    def __init__(self):
        # Cache the schema at initialization to avoid recomputation
        self._input_schema = {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Optional relative path to list files from. Defaults to current directory if not provided."
                }
            },
            "required": []
        }

    def name(self) -> str:
        """Return the tool name."""
        return "list_files"

    def description(self) -> str:
        """Return the tool description."""
        return "List files and directories at a given path. If no path is provided, lists files in the current directory."

    def input_schema(self) -> Dict[str, Any]:
        """Return the cached JSON schema for tool input validation."""
        return self._input_schema

    def execute(self, path: Optional[str] = None) -> str:
        """
        Execute the list files operation.

        Args:
            path: Optional relative path to list files from. Defaults to current directory.

        Returns:
            JSON string containing the list of files and directories

        Raises:
            Exception: If the directory cannot be listed
        """
        try:
            target_dir = path if path else "."

            files = []
            for root, dirs, filenames in os.walk(target_dir):
                # Get relative path from target directory
                rel_root = os.path.relpath(root, target_dir)

                # Add directories
                for dir_name in dirs:
                    if rel_root == ".":
                        files.append(f"{dir_name}/")
                    else:
                        files.append(f"{rel_root}/{dir_name}/")

                # Add files
                for filename in filenames:
                    if rel_root == ".":
                        files.append(filename)
                    else:
                        files.append(f"{rel_root}/{filename}")

            # Sort for consistent output
            files.sort()

            return json.dumps(files)

        except Exception as e:
            raise Exception(f"Error listing files in {path or '.'}: {str(e)}")

File: golang/tools/list_files.go
package tools

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/anthropics/anthropic-sdk-go"
)

// ListFilesTool implements the list_files functionality
type ListFilesTool struct {
	inputSchema anthropic.ToolInputSchemaParam
}

// ListFilesInput represents the input parameters for the list_files tool
type ListFilesInput struct {
	Path string `json:"path,omitempty" jsonschema_description:"Optional relative path to list files from. Defaults to current directory if not provided."`
}

// NewListFilesTool creates a new instance of the list files tool
func NewListFilesTool() *ListFilesTool {
	return &ListFilesTool{
		inputSchema: GenerateSchema[ListFilesInput](),
	}
}

// Name returns the tool name
func (t *ListFilesTool) Name() string {
	return "list_files"
}

// Description returns the tool description
func (t *ListFilesTool) Description() string {
	return "List files and directories at a given path. If no path is provided, lists files in the current directory."
}

// InputSchema returns the JSON schema for tool input validation
func (t *ListFilesTool) InputSchema() anthropic.ToolInputSchemaParam {
	return t.inputSchema
}

// Execute lists all files and directories in the specified path
func (t *ListFilesTool) Execute(input json.RawMessage) (string, error) {
	var listFilesInput ListFilesInput
	err := json.Unmarshal(input, &listFilesInput)
	if err != nil {
		return "", err
	}

	dir := "."
	if listFilesInput.Path != "" {
		dir = listFilesInput.Path
	}

	var files []string
	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		relPath, err := filepath.Rel(dir, path)
		if err != nil {
			return err
		}

		if relPath != "." {
			if info.IsDir() {
				files = append(files, relPath+"/")
			} else {
				files = append(files, relPath)
			}
		}
		return nil
	})

	if err != nil {
		return "", err
	}

	result, err := json.Marshal(files)
	if err != nil {
		return "", err
	}

	return string(result), nil
}
